### PR TITLE
Add `#[derive(FromRequestParts)]`

### LINKED ISSUE #1290: Add `#[derive(FromRequestParts)]` macro
### Motivation

I'm pretty sure many current users of the `FromRequest` macro are using the resulting type as a non-terminal extractor. After #1272, this will no longer be supported.

### Proposal

Add `FromRequestParts` macro after #1272, but before the 0.6 release.

### PR DESCRIPTION
**Pull request by jplatte**
Resolves #1290.

### CODE REVIEW

**Comment by davidpdrsn** [comment]
Looks good so far! Should we also add a couple of tests? I don't think we have to duplicate everything from `FromRequest` but should probably have a few smoke tests.

I think using `#[from_request]` for the attribute is fine. Not sure we should switch to `#[axum]` but if thats more common I guess it makes sense. I mostly use serde which uses `#[serde]` and derive-builder which `#[builder]`.

Might be a small risk of having multiple derives from axum on a single type which might conflict if they both used `#[axum]`. I don't think thats an issue we have now and probably isn't much risk.

**Comment by davidpdrsn** [comment]
I found a couple of edge cases so ended up duplicating all the "pass" UI tests just to be sure it works. Thats why the diff is so big.

I'll write some docs next.

**Comment by davidpdrsn** [comment]
@jplatte This is ready for review now. The code isn't super pretty but it works 😅

image

Not sure you can actually approve it since you opened the PR though.

File: axum-macros/src/from_request.rs [hunk 1]
```diff
@@ -323,58 +323,111 @@ fn extract_fields(
...
-                            .map(#into_inner);
-                        req = ::axum::http::Request::from_parts(parts, body);
-                        value
-                    },
-                })
+                let tokens = match tr {
+                    Trait::FromRequest => {
+                        quote_spanned! {ty_span=>
+                            #member: {
+                                let (mut parts, body) = req.into_parts();
```
[the remark this answers is not in the document]
Comment (jplatte) [reply axum-macros/src/from_request.rs:330]: If that's refactored, we may not need the branching here anymore.

File: axum-macros/src/lib.rs [hunk 2]
```diff
@@ -353,7 +356,53 @@ mod with_position;
...
+/// ```compile_fail
+/// use axum_macros::FromRequestParts;
+///
+/// #[derive(FromRequestParts)]
+/// struct MyExtractor {
+///     body: String,
+/// }
+/// ```
+///
+/// Use `#[derive(FromRequest)]` instead.
```
Comment (jplatte) [inline axum-macros/src/lib.rs:400]:
```suggestion
/// Use `#[derive(FromRequest)]` for that.
```

**Review by jplatte — COMMENTED** [review]
I approve 😄

File: axum-macros/src/from_request.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidpdrsn) [reply axum-macros/src/from_request.rs:330]: Yeah would be nice to refactor!

**Review by davidpdrsn — APPROVED** [review]
image
